Add VRF proof verification / generation to the Schnorrkel wrapper

David-Petrov opened the pull request.

Commit:
feat!: add vrf proof generation / verification

Commit:
feat: add test with invalid proof

Commit:
fix: fix error in big int to u64-le conversion

Commit:
fix: remove deprecated `jar` task

turns out, this task was left behind from the original source as a single-target build subtask to automatically detect the target currently being built for, but our fork targets multi-platform users, so we need all the target architecture builds

David-Petrov commented:
[KNOWN ISSUE] Build automation is broken. The original `jar` task for managing the native rust lib (removed by [015023e](https://github.com/LimeChain/polkaj/pull/2/commits/015023eec6e683886c6fda2596186904461c3f4c)) was meant to pick up the lib as output from cargo (built for the current architecture) and place it in its corresponding place. This, however, is a single-target automation.
Our fork of PolkaJ requires multi-target builds, so this approach is not suitable, thus `jar` has been removed. 
For now, manual intervention is necessary: paste the rust lib manually from `./polkaj-schnorrkel/build/rust/release` (that's the cargo build result; path relative from repository root) into its corresponding target architecture folder under `./polkaj-schnorrkel/src/main/resources/native`, before then building again to obtain the properly updated `.jar`.

Unfortunately, compilation for all target architectures must happen manually in a similar way as described above, before officially "deploying a new version" for multi-arch usage.

Is it worth investing effort into polishing the automation...?

[RESOLVED]: Discussed with @georg-getz, polishing the automation is not worth it for now, as priorities are different. This goes into `TODO.md`. For now, a detailed description of this necessary manual intervention is to be added to this fork's `README.md`.

Commit:
chore!: remove a leaked playground main

whoops

David-Petrov commented on polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs:
```diff
@@ -0,0 +1,54 @@
+      //  Figure out how to make this simply a `Box<[u8]>` instead of a `Vec<Box<[u8]>>`,
+      //  currently a blocker with byte[] fields (seems to be an issue with robusta itself, some missing impls?)
+      #[field]
+      pub domainSeparationLabel: Field<'env, 'borrow, Vec<Box<[u8]>>>,
```
That's a patch fix. I mean the field being a `Vec<Box<[u8]>>`, which corresponds to an `ArrayList<byte[]>` instead of simply a `Box<[u8]>`, which corresponds to a `byte[]`. Some weird, deeper rooted trait-resolution issues arise, which (I suspect) is a bug with robusta itself... Possible solutions:
1. Fix robusta (I've tried, but not to much avail).
2. Throw robusta away and couple ourselves to primitive `jni` signatures, thus either manually mimick the `TryFrom/IntoJavaValue` traits for more complex objects or simply expand signatures to unpack the inner state representations (and end up with a lot of function arguments...). The former will require work, the latter is lazy but scales really badly.
3. Leave robusta in for its good parts - mainly the `TryFrom/IntoJavaValue` traits used for now - but don't try to derive them and instead manually implement them (I'm not yet sure how feasible this is).

This issue also directly affects the decision about the (for now) ugly return strategy on the `_vrfSign` native function.

David-Petrov replied:
After considerable headaches, all three approaches turn out to be a bigger headache than expected. Thus, for now the hacks remain as-is (I deemed them not that big of a pain), an attempt at a proper documentation of the decisions made has been made.

Our biggest hope lies with the active participation in upgrading robusta to jni to 0.21, so we're following those PRs: https://github.com/giovanniberti/robusta/pull/50 (and https://github.com/giovanniberti/robusta/pull/59 related).

May need later revisiting.

georg-getz commented on polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/schnorrkel/VrfOutputAndProof.java:
```diff
@@ -0,0 +1,40 @@
+        this.proof = proof;
+    }
+
+    public byte[] getOutput() {
+        return output;
+    }
+
+    public byte[] getProof() {
+        return proof;
+    }
```
Remove and add @Getter annotation.

David-Petrov replied:
No dependency (e.g. Lombok) is currently included that provides this, not worth it IMO for now.

georg-getz commented on polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/schnorrkel/VrfOutputAndProof.java:
```diff
@@ -0,0 +1,40 @@
+ * Essentially a record to hold the VRF output together with its proof.
+ */
+public class VrfOutputAndProof {
+    public static int OUTPUT_BYTE_LEN = 32;
+    public static int PROOF_BYTE_LEN = 64;
```
add `final`

Commit:
chore: add `final`

David-Petrov commented on polkaj-schnorrkel/src/main/java/io/emeraldpay/polkaj/merlin/TranscriptData.java:
```diff
@@ -0,0 +1,65 @@
+     * @param label the message label
+     * @param x the u64 to append
+     */
+    // TODO:
+    //  Maybe just remove this method and leave it up to the user...
+    //  We can't guarantee good enough type safety and obeying the representational contract is left to the developer,
+    //  which could lead to issues.
```
Self explanatory. Still, leaving this convenience method in the public interface of this here `TranscriptData` (inspired by the original rust `schnorrkel`) might not be such a good idea in the Java world, since we don't have an unified `u64` type, so now it's up to the caller to understand my `NOTE:` in the javadoc... which stinks.

David-Petrov replied:
I've removed it. We'll manage this u64 representation as a `byte[]` from Fruzhin.

David-Petrov commented on polkaj-schnorrkel/src/rust/src/merlin_jni/mod.rs:
```diff
@@ -0,0 +1,54 @@
+  }
+}
+
+// TODO:
+//  Try to avoid having to `Box::leak`,
+//  In theory, should be safe enough if used correctly Java-side, since "we know Java owns the object"
```
That's also a topic I'm rather unsure about. Leaking the memory (effectively telling rust "leave it up to me to handle") should be okay since the memory is actually owned and managed by the Java runtime. I am, however, lacking the deepest understanding as of what possible issues we could have with that type of memory management :D

David-Petrov replied:
After digging into this for a little bit, I managed to convince myself that's rather OK. Proper documentation has been added on the intentionality of this piece of code.

David-Petrov commented on polkaj-schnorrkel/src/main/resources/native/macos/arm/libpolkaj_schnorrkel.dylib:
The removal of this file has to do with forcing recompilation of the native rust lib, since automatic rehydration was broken. Will have to be manually added again (with the finalised changes on the rust side) directly before merging this PR... Or we fix the automations.

David-Petrov replied:
All has been manually rebuilt.

Commit:
refactor!: remove `appendU64` from `TranscriptData`

dubbed too inconvenient as a public API method since java doesn't provide a u64 type, so the correct representation is left to the caller

Commit:
refactor!: proper error handling

add a sensible differentiation between "false" as "incorrect proof" and when an exception must be thrown in Java

Commit:
chore: better document hacks

add proper explanations for weird decisions in the respective places

Commit:
meta: add .vscode

useful to link rust source to a cargo.toml for VSCode users

Commit:
chore: use `merlin` with `std`

no need to explicitly disable it as a default feature... for now

Commit:
chore: add `Cargo.lock` to version control

Commit:
feat!: rebuild the native lib for all target architectures

Commit:
chore: bump minor version

Commit:
chore: link tests to where they've been translated from

Commit:
style: unify style

- remove semicolons;
- simplify boolean expressions
